lib/puppet/pops/serialization/serializer.rb

```diff
@@ -0,0 +1,91 @@
+require_relative 'extension'
+
+module Puppet::Pops
+module Serialization
+  # The serializer is capable of writing, arrays, maps, and complex objects using an underlying protocol writer. It takes care of
+  # tabulating and disassembling complex objects.
+  # @api public
+  class Serializer
+    # @param [AbstractWriter] writer the writer that is used for writing primitive values
+    # @api public
+    def initialize(writer)
+      @written = {}
+      @writer = writer
+    end
+
+    # Tell the underlying writer to finish
+    # @api public
+    def finish
+      @writer.finish
+    end
+
+    # Write an object
+    # @param [Object] value the object to write
+    # @api public
+    def write(value)
+      case value
+      when Integer, Float, String, true, false, nil, Time
```

Comment:
There are many problems with Integer. Both unsigned 64 bits, and bigints (>64 bit) in most serialization formats. Also, in puppet, these larger values do not work throughout the infrastructure. Hence, we cannot allow any Ruby Integer. On 32 bit platforms (IIRC) FixNum is not full 64 bit, and MRI switches to BigInt when overflowing 32 bits. Therefore it is not just a matter of changing the Integer to a FixNum (as it will be too restrictive on some platforms).  I think Pcore should have signed Integer 64 bit as max for integers and that it is enforced by serialization. We can have another ticket to add support for BigInt in Pcore (a separate data type).